File: Mex/SDOF_FreeResponse/Fortran_MEX/MEXFUNCTION.f90
#include <fintrf.h>

SUBROUTINE MEXFUNCTION(NLHS, PLHS, NRHS, PRHS)
    ! This subroutine is the main gateway to MATLAB.  When a MEX function
    !  is executed MATLAB calls this subroutine

    USE, INTRINSIC :: ISO_C_BINDING, ONLY: C_SIZE_T
#if defined(_WIN32) || defined(_WIN64)
    USE IFPORT
#endif
    USE MX_INTERFACES
    USE MEX_INTERFACES

    IMPLICIT NONE
#if defined(_WIN32) || defined(_WIN64)
!DEC$ ATTRIBUTES DLLEXPORT :: MEXFUNCTION
#endif
    MWPOINTER PLHS(*), PRHS(*)
    INTEGER(4) NLHS, NRHS
    REAL(8) w_n,zeta,x0,v0
    mwSize N
    INTEGER(C_SIZE_T) NN
    REAL(8), ALLOCATABLE :: t(:),x(:)

    INTERFACE
        FUNCTION FREE_RESPONSE(w_n,zeta,x0,v0,t,N)
            USE, INTRINSIC :: ISO_C_BINDING, ONLY: C_SIZE_T
            INTEGER(C_SIZE_T), INTENT(IN) :: N
            REAL(8) w_n,zeta,x0,v0,t(N),FREE_RESPONSE(N)
        END FUNCTION FREE_RESPONSE
    END INTERFACE

#if defined(_WIN32) || defined(_WIN64)
! For Windows only!
! This resets the floating point exception to allow divide by zero,
! overflow and invalid numbers. 
!
    INTEGER(2) CONTROL
    CALL GETCONTROLFPQQ(CONTROL)
    CONTROL = CONTROL .OR. FPCW$ZERODIVIDE
        CONTROL = CONTROL .OR. FPCW$INVALID
        CONTROL = CONTROL .OR. FPCW$OVERFLOW
    CALL SETCONTROLFPQQ(CONTROL)
#endif

    ! Check for proper number of arguments
    IF ((NRHS/=5)) THEN
        CALL mexErrMsgTxt('Dear student, This function needs 5 inputs.')
    ENDIF

    IF ((NLHS>1)) THEN
        CALL mexErrMsgTxt('Dear student, this function returns only one output.')
    ENDIF

    DO N=1,4
        IF (mxGetNumberOfElements(PRHS(N)) /= 1) THEN
            CALL mexErrMsgTxt('Dear student, first four inputs must be scalars.')    !Improve this
        ENDIF
    END DO

    w_n=mxGetScalar(PRHS(1))
    zeta=mxGetScalar(PRHS(2))
    x0=mxGetScalar(PRHS(3))
    v0=mxGetScalar(PRHS(4))
    N=mxGetNumberOfElements(PRHS(5))
    ALLOCATE(t(N),x(N))
    CALL mxCopyPtrToReal8(mxGetDoubles(PRHS(5)), t, N)

    NN=N
    x=FREE_RESPONSE(w_n,zeta,x0,v0,t,NN)

    PLHS(1)=mxCreateDoubleMatrix(mxGetM(PRHS(5)),mxGetN(PRHS(5)),INT(0,4))
    CALL mxCopyReal8ToPtr(x, mxGetDoubles(PLHS(1)), N)

END SUBROUTINE MEXFUNCTION

File: Mex/SDOF_FreeResponse/Cpp_MEX/mexFunction.cpp
#include <valarray>
#include "mex.hpp"
#include "mexAdapter.hpp"

using namespace matlab::data;
using matlab::mex::ArgumentList;

std::valarray<double> Free_Response(const double& w_n, const double& zeta, const double& x0, const double& v0, const std::valarray<double>& t);

class MexFunction : public matlab::mex::Function
{
	ArrayFactory factory;
	
public:
	void operator()(ArgumentList outputs, ArgumentList inputs)
	{
		checkArguments(outputs, inputs);
		
		const double w_n=inputs[0][0];
		const double zeta= inputs[1][0];
		const double x0= inputs[2][0];
		const double v0= inputs[3][0];
		const size_t N= inputs[4].getNumberOfElements();
		std::valarray<double> t(N);
		for (size_t n = 0; n < N; n++)
			t[n]= inputs[4][n];
	
		// Do the actual computations
		std::valarray<double> x=Free_Response(w_n,zeta,x0,v0,t);

		TypedArray<double> x_out=
		outputs[0] = factory.createArray<double>(inputs[4].getDimensions(), &x[0], &x[N]);

		return;
	}

	void checkArguments(ArgumentList outputs, ArgumentList inputs)
	{
		const std::shared_ptr<matlab::engine::MATLABEngine> matlabPtr = getEngine();

		if (inputs.size() != 5)
			matlabPtr->feval(u"error", 0,
				std::vector<Array>({ factory.createScalar("Dear student, This function needs 5 inputs.") }));

		if (outputs.size() > 1)
			matlabPtr->feval(u"error", 0,
				std::vector<Array>({ factory.createScalar("Dear student, this function returns only one output.") }));

		for (size_t n = 0; n < 4; n++)
		{
			if (inputs[n].getNumberOfElements() != 1)
				matlabPtr->feval(u"error", 0,
					std::vector<Array>({ factory.createScalar("Dear student, first four inputs must be scalars.") }));    //Improve this
		}
	};
};